File: front/src/types/competition.ts
export type Competition = {
  id: number;
  url: string;
  type: string | null;
  name: string | null;
  organizer: string | null;
  deadline: string | null;
  deadline_date: string | null;
  event_date: string | null;
  event_date_date: string | null;
  event_end_date: string | null;
  venue: string | null;
  prize: string | null;
  prize_amount: number | null;
  target: string | null;
  difficulty: string | null;
  team_size_min: number | null;
  team_size_max: number | null;
  description: string | null;
  score: number | null;
};

export type CompetitionList = {
  fetched_at: string;
  count: number;
  items: Competition[];
};

export type CompetitionSearch = {
  fetched_at: string;
  query: string;
  interpretation: string;
  count: number;
  results: Competition[];
};

export type EntryStatus = "challenge" | "wait" | "achieve" | "dropped";

export type Entry = {
  id: string;
  user_id: string;
  competition_id: number | null;
  url: string;
  name: string;
  status: EntryStatus;
  memo: string | null;
  result: string | null;
  deadline_date: string | null;
  event_date_date: string | null;
  activity_id: string | null;
  applied_at: string;
  decided_at: string | null;
  username: string | null;
  display_name: string | null;
  avatar_url: string | null;
};

export type InternalEvent = {
  id: string;
  name: string;
  event_date: string;
  event_end_date: string | null;
  venue: string | null;
  description: string | null;
};

export type DashboardCard = {
  card_key: string;
  visible: boolean;
  order: number;
};

export type DashboardSummary = {
  entries: {
    challenge: number;
    wait: number;
    achieve: number;
    dropped: number;
  };
  upcoming: {
    entry_id: string;
    name: string;
    url: string;
    deadline_date: string;
    days_left: number;
  }[];
  week: { date: string; minutes: number }[];
  cards: DashboardCard[];
};

export type Portfolio = {
  user_id: string;
  discord_id: string;
  username: string;
  display_name: string | null;
  avatar_url: string | null;
  bio: string | null;
  business_desc: string | null;
  sns_links: Record<string, string> | null;
  public: boolean;
  total_points: number;
  total_hours: number;
  achievement_count: number;
  skills: string[];
  items: {
    date: string;
    kind: "activity" | "achievement";
    title: string;
    body: string | null;
    points: number | null;
    url: string | null;
  }[];
};

export const COMPETITION_TYPES = [
  { value: "hackathon", label: "ハッカソン" },
  { value: "bizcon", label: "ビジコン" },
  { value: "academia", label: "学術" },
  { value: "startup", label: "スタートアップ" },
  { value: "acceleration", label: "アクセラ" },
  { value: "networking", label: "交流" },
] as const;

export const ENTRY_STATUS_LABELS: Record<EntryStatus, string> = {
  challenge: "応募",
  wait: "結果待ち",
  achieve: "成果",
  dropped: "見送り",
};

File: back/app/schemas/competition.py
from datetime import date, datetime
from typing import Any, Optional

from pydantic import BaseModel, Field


class CompetitionListResponse(BaseModel):
    """外部APIのコンペ一覧 + キャッシュ鮮度。"""

    fetched_at: datetime
    count: int
    items: list[dict[str, Any]]


class CompetitionSearchResponse(BaseModel):
    fetched_at: datetime
    query: str
    interpretation: str
    count: int
    results: list[dict[str, Any]]


# --- 応募トラッキング ---


class EntryCreate(BaseModel):
    name: str
    # 外部コンペのURL。学内表彰など URL が無い成果もあるので任意
    url: str = ""
    competition_id: Optional[int] = None
    memo: Optional[str] = None
    deadline_date: Optional[str] = None
    event_date_date: Optional[str] = None
    # "challenge"（通常の応募）か "achieve"（過去の受賞をまとめて登録する場合）
    status: str = "challenge"
    result: Optional[str] = None


class EntryUpdate(BaseModel):
    status: Optional[str] = None
    memo: Optional[str] = None
    result: Optional[str] = None


class EntryResponse(BaseModel):
    model_config = {"from_attributes": True}

    id: str
    user_id: str
    competition_id: Optional[int]
    url: str
    name: str
    status: str
    # memo は本人・管理者以外には None にして返す
    memo: Optional[str]
    result: Optional[str]
    deadline_date: Optional[str]
    event_date_date: Optional[str]
    activity_id: Optional[str]
    applied_at: datetime
    decided_at: Optional[datetime]
    # 成果にならなかった場合に自動削除されるまでの残り日数（None = 期日不明で自動削除しない）
    expires_in_days: Optional[int] = None
    # 表示用（一覧で誰の応募かを出す）
    username: Optional[str] = None
    display_name: Optional[str] = None
    avatar_url: Optional[str] = None


# --- 自団体イベント ---


class InternalEventCreate(BaseModel):
    name: str
    event_date: date
    event_end_date: Optional[date] = None
    venue: Optional[str] = None
    description: Optional[str] = None


class InternalEventResponse(BaseModel):
    model_config = {"from_attributes": True}

    id: str
    name: str
    event_date: date
    event_end_date: Optional[date]
    venue: Optional[str]
    description: Optional[str]
    created_at: datetime


# --- ダッシュボード設定 ---


class DashboardPrefItem(BaseModel):
    card_key: str
    visible: bool = True
    order: int = Field(default=0, ge=0)


class DashboardPrefUpdate(BaseModel):
    cards: list[DashboardPrefItem]


# --- スキルタグ ---


class SkillCreate(BaseModel):
    label: str = Field(min_length=1, max_length=40)


class SkillResponse(BaseModel):
    model_config = {"from_attributes": True}

    id: str
    label: str
    source: str
